aten/src/ATen/native/cuda/Randperm.cu

```diff
@@ -47,7 +47,10 @@ template <int N> struct alignas(N) OpaqueType { char data[N]; };
 
 Tensor& randperm_out_cuda(int64_t n, c10::optional<Generator> generator, Tensor& result) {
   TORCH_CHECK(n >= 0, "n must be non-negative, got", n);
-  TORCH_CHECK(!generator.has_value() || (generator.has_value() && result.device() == generator->device()), "Expected a '", result.device(), "' generator device but found '", generator->device(), "'");
+  TORCH_CHECK(!generator.has_value() || (result.device().type() == generator->device().type()),
```

Comment:
please use 
```
auto generator = get_generator_or_default<CUDAGeneratorImpl>(gen_, cuda::detail::getDefaultCUDAGenerator());
```
like in all other places. 
This would not error out for fully specified target and rng devices, but this is the case for other calls now anyway:
```
n [6]: import torch
   ...: rng = torch.Generator(device="cuda:1")
   ...: idx_pts = torch.randperm(n=50, generator=rng, device="cuda:1")

In [7]: torch.randint(low=0, high=10, size=(1,), generator=rng, device="cuda")
Out[7]: tensor([5], device='cuda:0')

In [8]: torch.randint(low=0, high=10, size=(1,), generator=rng, device="cuda:0")
Out[8]: tensor([6], device='cuda:0')
```

Reply:
thanks for the advice. I was looking at the places below that are using the original `c10::optional<Generator> generator`

- `_random(..., c10::optional<Generator>)`
- `randperm_handle_duplicate_keys(..., c10::optional<Generator>)`

These two require `c10::optional<Generator>` not `CUDAGeneratorImpl*`. Is there a way to convert it back to the type `c10::optional<Generator>`? 

Otherwise, `get_generator_or_default` only replaces the `generator.has_value()` in the TORCH_CHECK, because at least the generator and the result need to have the same device type. 


----

Also, to be clear, do we want this to throw exception (looks like `torch.randint` not error out)?

```
torch.randperm(n=50, generator=torch.Generator(device="cuda:1"), device="cuda:0")
```

Reply:
To be consistent with other random functions, we don't want this to throw exception. Longer term, generator should ignore device ordinal, since it's not used anyway. 
Sorry, I don't know how to convert CUDAGeneratorImpl* to c10::optional<Generator>. Could you not check anything, and rely on `random_` to throw an error if the generator is CPU generator?

Reply:
@ngimel thanks, makes sense, updated, and now just let `random_` handles it. Also updated the tests' location to better reflect the edge cases.